crates/ruff/src/rules/refurb/rules/reimplemented_starmap.rs

```diff
@@ -0,0 +1,304 @@
+use ruff_diagnostics::{AutofixKind, Diagnostic, Edit, Fix, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Expr};
+use ruff_python_codegen::Generator;
+use ruff_python_semantic::SemanticModel;
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Checks for generator expressions, list and set comprehensions that can
+/// be replaced with `itertools.starmap`.
+///
+/// ## Why is this bad?
+/// When unpacking values from iterators to pass them directly to
+/// a function, prefer `itertools.starmap`.
+///
+/// Using `itertools.starmap` is more concise and readable.
+///
+/// ## Example
+/// ```python
+/// scores = [85, 100, 60]
+/// passing_scores = [60, 80, 70]
+///
+///
+/// def passed_test(score: int, passing_score: int) -> bool:
+///     return score >= passing_score
+///
+///
+/// passed_all_tests = all(
+///     passed_test(score, passing_score)
+///     for score, passing_score in zip(scores, passing_scores)
+/// )
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from itertools import starmap
+///
+///
+/// scores = [85, 100, 60]
+/// passing_scores = [60, 80, 70]
+///
+///
+/// def passed_test(score: int, passing_score: int) -> bool:
+///     return score >= passing_score
+///
+///
+/// passed_all_tests = all(starmap(passed_test, zip(scores, passing_scores)))
+/// ```
+///
+/// ## References
+/// - [Python documentation: `itertools.starmap`](https://docs.python.org/3/library/itertools.html#itertools.starmap)
+#[violation]
+pub struct ReimplementedStarmap;
+
+impl Violation for ReimplementedStarmap {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use `itertools.starmap` instead of the generator")
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        Some(format!("Replace with `itertools.starmap`"))
+    }
+}
+
+/// A abstract node that can be considered a candidate for replacement with `starmap`.
+pub(crate) trait StarmapEquivalent {
+    /// Get generated element.
+    fn get_element(&self) -> &Expr;
+    /// Get generator comprehensions.
+    fn get_generators(&self) -> &[ast::Comprehension];
+    /// Get fix suggestion transforming this node into a call to `starmap`.
+    fn make_suggestion(name: String, iter: &Expr, func: &Expr, generator: Generator) -> String;
+    /// Check if it is possible to apply the fix in the current context.
+    fn check_fix_possibility(semantic: &SemanticModel) -> bool;
+}
+
+// FURB140
+pub(crate) fn reimplemented_starmap<T: StarmapEquivalent + Ranged>(
```

Comment:
The use of a generic here means that Rust generates a dedicated implementation for each `T` instantiation. 

I think it may be better to define a `AnyComprehension` enum (or similar) with a variant for each node that you want to support instead of using a trait.  See 

https://github.com/astral-sh/ruff/blob/c05e4628b1d385d106e7e3d355f5d1d76fbfe401/crates/ruff_python_formatter/src/expression/expr_yield.rs#L10-L35

Reply:
I'm not a huge fan of such a solution. With the trait, we don't need to perform any runtime checks for the given type. With the enum, we do them 4 times: `element`, `generators`, `range`, and `try_fix`. These seem not critical in any way, but one can argue that bloating from instantiating the same (rather small) generic 3 times is kind of similar.

I guess I can see how an enum solution is better for decoupling, if I remove `try_fix` from that interface altogether. Then it can be used in other places that share comprehensions. On the other hand, this interface can't include `DictComprehension` without making `element` optional and adding more runtime checks.

Sorry, for the rant. If you insist, I can change it, but I hope you see my point here 😄